ngrinder-controller/src/main/java/org/ngrinder/agent/service/AgentService.java

```diff
@@ -0,0 +1,489 @@
+	}
+
+
+	private String createAgentKey(String ip, String name) {
+		return ip + "_" + name;
+	}
+
+	/**
+	 * Ready agent state count return
+	 *
+	 * @param user The login user
+	 * @param targetRegion targetRegion The name of target region
+	 * @return ready Agent count
+	 */
+	@Override
+	public int getReadyAgentCount(User user, String targetRegion) {
+		int readyAgentCnt = 0;
+		String myOwnAgent = targetRegion + "_owned_" + user.getUserId();
+		for (AgentInfo agentInfo : getAllReady()) {
+			if (!agentInfo.isApproved()) {
+				continue;
+			}
+			String fullRegion = agentInfo.getRegion();
+			if (StringUtils.equals(fullRegion, targetRegion) || StringUtils.equals(fullRegion, myOwnAgent)) {
+				readyAgentCnt++;
+			}
+		}
+		return readyAgentCnt;
+	}
+
+	@Override
+	public List<AgentInfo> getAllVisible() {
+		return getAllActive();
+	}
+
+	@Override
+	public List<AgentInfo> getAllAttached() {
+		return agentInfoStore.getAllAgentInfo();
+	}
+
+	@Override
+	public List<AgentInfo> getLocalAgents() {
+		return getAllAttached();
+	}
+
+	@Override
+	public String createKey(AgentInfo agentInfo) {
+		return agentInfo.getAgentKey();
+	}
+
+	@Override
+	public String createKey(AgentControllerIdentityImplementation agentIdentity) {
+		return createAgentKey(agentIdentity.getIp(), agentIdentity.getName());
+	}
+
+	private void publishTopic(AgentInfo agentInfo, AgentRequest.RequestType requestType) {
+		hazelcastService.publish(AGENT_TOPIC_NAME, new TopicEvent<>(AGENT_TOPIC_LISTENER_NAME,
+			agentManager.extractRegionKey(agentInfo.getRegion()), new AgentRequest(agentInfo.getIp(), agentInfo.getName(), requestType)));
+	}
+
+	@Override
+	public void execute(TopicEvent<AgentRequest> event) {
+		if (event.getKey().equals(config.getRegion())) {
+			AgentRequest agentRequest = event.getData();
+			AgentControllerIdentityImplementation agentIdentity = getAgentIdentityByIpAndName(agentRequest.getAgentIp(), agentRequest.getAgentName());
+			if (agentIdentity != null) {
+				agentRequest.getRequestType().process(AgentService.this, agentIdentity);
+			}
+		}
+	}
+
+	@Override
+	public void update(Map<AgentIdentity, AgentProcessControlImplementation.AgentStatus> agentMap) {
+		boolean approved = config.getControllerProperties().getPropertyBoolean(PROP_CONTROLLER_ENABLE_AGENT_AUTO_APPROVAL);
+		Set<AgentInfo> agentInfoSet = agentInfoStore.getAllAgentInfo()
+			.stream()
+			.filter(agentInfo -> StringUtils.equals(agentManager.extractRegionKey(agentInfo.getRegion()), config.getRegion()))
+			.collect(toSet());
+
+		for (AgentProcessControlImplementation.AgentStatus status : agentMap.values()) {
+			AgentControllerIdentityImplementation agentIdentity = (AgentControllerIdentityImplementation) status.getAgentIdentity();
+			AgentInfo agentInfo = agentInfoStore.getAgentInfo(createKey(agentIdentity));
+			// check new agent
+			if (agentInfo == null) {
+				agentInfo = new AgentInfo();
+			}
+
+			fillUpAgentInfo(agentInfo, status);
```

Comment:
Do we need periodic fillup? listener isn't enought?

Reply:
This method called by `AgentStatusUpdateListener` only. not periodically called.

Reply:
okey